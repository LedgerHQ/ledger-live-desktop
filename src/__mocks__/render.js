import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { I18nextProvider } from 'react-i18next'

import createStore from 'renderer/createStore'
import i18n from 'renderer/i18n/electron'

import theme from 'styles/theme'

export default function render(component, state) {
  const store = createStore({ state })
  return renderer
    .create(
      <I18nextProvider i18n={i18n} initialLanguage="en">
        <Provider store={store}>
          <ThemeProvider theme={theme}>{component}</ThemeProvider>
        </Provider>
      </I18nextProvider>,
    )
    .toJSON()
}
