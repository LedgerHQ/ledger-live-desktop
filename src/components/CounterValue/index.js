// @flow

import type { BigNumber } from 'bignumber.js'
import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import type { Currency } from '@ledgerhq/live-common/lib/types'

import {
  counterValueCurrencySelector,
  exchangeSettingsForPairSelector,
  intermediaryCurrency,
} from 'reducers/settings'
import CounterValues from 'helpers/countervalues'

import FormattedVal from 'components/base/FormattedVal'

import type { State } from 'reducers'

type OwnProps = {|
  // wich market to query
  currency: Currency,

  // when? if not given: take latest
  date?: Date,

  value: BigNumber,

  alwaysShowSign?: boolean,

  subMagnitude?: number,

  placeholder?: React$Node,

  prefix?: React$Node,
  suffix?: React$Node,
|}

type StateProps = {|
  counterValueCurrency: Currency,
  value: ?number,
|}

type Props = {|
  ...OwnProps,
  ...StateProps,
|}

const mapStateToProps = (state: State, props: OwnProps) => {
  const { currency, value, date, subMagnitude } = props
  const counterValueCurrency = counterValueCurrencySelector(state)
  const intermediary = intermediaryCurrency(currency, counterValueCurrency)
  const fromExchange = exchangeSettingsForPairSelector(state, { from: currency, to: intermediary })
  const toExchange = exchangeSettingsForPairSelector(state, {
    from: intermediary,
    to: counterValueCurrency,
  })
  const counterValue = CounterValues.calculateWithIntermediarySelector(state, {
    from: currency,
    fromExchange,
    intermediary,
    toExchange,
    to: counterValueCurrency,
    value,
    date,
    disableRounding: !!subMagnitude,
  })

  return {
    counterValueCurrency,
    value: counterValue,
  }
}

class CounterValue extends PureComponent<Props> {
  static defaultProps = {
    alwaysShowSign: true, // FIXME this shouldn't be true by default
  }

  render() {
    const {
      value,
      counterValueCurrency,
      date,
      alwaysShowSign,
      placeholder,
      prefix,
      suffix,
      ...props
    } = this.props

    if (!value) {
      return placeholder || null
    }

    return (
      <>
        {prefix || null}
        <FormattedVal
          val={value}
          unit={counterValueCurrency.units[0]}
          showCode
          alwaysShowSign={alwaysShowSign}
          {...props}
        />
        {suffix || null}
      </>
    )
  }
}

export default connect(mapStateToProps)(CounterValue)
