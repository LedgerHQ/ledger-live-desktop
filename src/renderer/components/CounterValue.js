// @flow
import { BigNumber } from "bignumber.js";
import React from "react";
import { useSelector } from "react-redux";
import type { Currency } from "@ledgerhq/live-common/lib/types";
import { useCalculate } from "@ledgerhq/live-common/lib/countervalues/react";
import { counterValueCurrencySelector } from "~/renderer/reducers/settings";
import FormattedVal from "~/renderer/components/FormattedVal";

type Props = {
  // wich market to query
  currency: Currency,

  // when? if not given: take latest
  date?: Date,

  value: BigNumber,

  alwaysShowSign?: boolean,

  subMagnitude?: number,

  placeholder?: React$Node,

  prefix?: React$Node,
  suffix?: React$Node,
};

export default function CounterValue({
  value,
  date,
  currency,
  alwaysShowSign = false,
  placeholder,
  prefix,
  suffix,
  ...props
}: Props) {
  const counterValueCurrency = useSelector(counterValueCurrencySelector);
  const countervalue = useCalculate({
    from: currency,
    to: counterValueCurrency,
    value: value.toNumber(),
    disableRounding: true,
  });

  if (typeof countervalue === "undefined") {
    return placeholder || null;
  }

  return (
    <>
      {prefix || null}
      <FormattedVal
        {...props}
        val={BigNumber(countervalue)}
        currency={currency}
        unit={counterValueCurrency.units[0]}
        showCode
        alwaysShowSign={alwaysShowSign}
      />
      {suffix || null}
    </>
  );
}
