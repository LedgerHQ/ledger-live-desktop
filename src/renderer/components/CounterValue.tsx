import { BigNumber } from "bignumber.js";
import React, { useEffect, useMemo } from "react";
import { useSelector } from "react-redux";
import { Currency } from "@ledgerhq/live-common/lib/types";
import {
  useCalculate,
  useCountervaluesPolling,
} from "@ledgerhq/live-common/lib/countervalues/react";
import { counterValueCurrencySelector } from "~/renderer/reducers/settings";
import FormattedVal from "~/renderer/components/FormattedVal";
import ToolTip from "./Tooltip";
import { Trans } from "react-i18next";
import useTheme from "~/renderer/hooks/useTheme";
import { addExtraSessionTrackingPair, useTrackingPairs } from "../actions/general";

type Props = {
  // wich market to query
  currency: Currency;

  // when? if not given: take latest
  date?: Date;

  value: BigNumber | number;

  alwaysShowSign?: boolean;
  alwaysShowValue?: boolean; // overrides discreet mode

  subMagnitude?: number;

  placeholder?: React$Node;

  prefix?: React$Node;
  suffix?: React$Node;
  placeholderStyle?: { [key: string]: string | number };
  color?: string;
};

export const NoCountervaluePlaceholder = ({
  placeholder,
  style = {},
}: {
  placeholder?: React$Node;
  style?: any;
}) => {
  const colors = useTheme("colors");

  return (
    <div style={{ ...style, maxHeight: "16px" }}>
      <ToolTip
        content={<Trans i18nKey="errors.countervaluesUnavailable.title" />}
        containerStyle={{ color: colors.palette.text.shade40 }}
      >
        {placeholder || "-"}
      </ToolTip>
    </div>
  );
};

export default function CounterValue({
  value: valueProp,
  date,
  currency,
  alwaysShowSign = false,
  alwaysShowValue = false,
  placeholder,
  prefix,
  suffix,
  placeholderStyle,
  ...props
}: Props) {
  const value = valueProp instanceof BigNumber ? valueProp.toNumber() : valueProp;
  const counterValueCurrency = useSelector(counterValueCurrencySelector);
  const trackingPairs = useTrackingPairs();
  const cvPolling = useCountervaluesPolling();
  const hasTrackingPair = useMemo(
    () => trackingPairs.some(tp => tp.from === currency && tp.to === counterValueCurrency),
    [counterValueCurrency, currency, trackingPairs],
  );

  useEffect(() => {
    let t;
    if (!hasTrackingPair) {
      addExtraSessionTrackingPair({ from: currency, to: counterValueCurrency });
      t = setTimeout(cvPolling.poll, 2000); // poll after 2s to ensure debounced CV userSettings are effective after this update
    }

    return () => {
      if (t) clearTimeout(t);
    };
  }, [counterValueCurrency, currency, cvPolling, cvPolling.poll, hasTrackingPair, trackingPairs]);

  const countervalue = useCalculate({
    from: currency,
    to: counterValueCurrency,
    value,
    disableRounding: true,
    date,
  });

  if (typeof countervalue !== "number") {
    return <NoCountervaluePlaceholder placeholder={placeholder} style={placeholderStyle} />;
  }

  return (
    <>
      {prefix || null}
      <FormattedVal
        {...props}
        val={countervalue}
        currency={currency}
        unit={counterValueCurrency.units[0]}
        showCode
        alwaysShowSign={alwaysShowSign}
        alwaysShowValue={alwaysShowValue}
      />
      {suffix || null}
    </>
  );
}
