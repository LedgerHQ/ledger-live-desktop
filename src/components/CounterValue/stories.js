// @flow

import React from 'react'
import { getCryptoCurrencyById } from '@ledgerhq/live-common/lib/helpers/currencies'
import { storiesOf } from '@storybook/react'
import { number } from '@storybook/addon-knobs'

import CounterValue from 'components/CounterValue'

const stories = storiesOf('Components', module)

const currency = getCryptoCurrencyById('bitcoin')

stories.add('CounterValue', () => (
  <CounterValue exchange="KRAKEN" currency={currency} value={number('value', 100000000)} />
))
