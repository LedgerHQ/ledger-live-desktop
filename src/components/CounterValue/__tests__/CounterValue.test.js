// @flow

import React from 'react'

import render from '__mocks__/render'
import CounterValue from '..'

describe('components', () => {
  describe('CounterValue', () => {
    it('basic', () => {
      const state = { counterValues: { BTC: { USD: { latest: 10e2 } } } }
      const component = <CounterValue ticker="BTC" value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('specifying ticker different from default', () => {
      const state = { counterValues: { LOL: { USD: { latest: 5e2 } } } }
      const component = <CounterValue ticker="LOL" value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('using countervalue different from default', () => {
      const state = {
        counterValues: { BTC: { EUR: { latest: 42 } } },
        settings: {
          counterValue: 'EUR',
        },
      }
      const component = <CounterValue ticker="BTC" value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('without countervalues populated', () => {
      const state = { counterValues: {} }
      const component = <CounterValue ticker="BTC" value={1} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('with time travel whith date in countervalues', () => {
      const state = { counterValues: { BTC: { USD: { '2018-01-01': 20e2 } } } }

      const date = new Date('2018-01-01')
      const component = <CounterValue ticker="BTC" value={1} date={date} />
      const tree = render(component, state)
      expect(tree).toMatchSnapshot()
    })

    it('with time travel whith date not in countervalues', () => {
      const state = { counterValues: { BTC: { USD: { '2018-01-01': 20e2 } } } }
      const date = new Date('2018-01-02')
      const component = <CounterValue ticker="BTC" value={1} date={date} />
      const tree = render(component, state)

      // TODO: actually it returns 0 when date is not in countervalues
      // do we want to use closest past value instead?
      expect(tree).toMatchSnapshot()
    })
  })
})
