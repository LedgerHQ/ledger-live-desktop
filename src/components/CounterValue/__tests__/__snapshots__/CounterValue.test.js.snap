// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CounterValue basic 1`] = `
<div
  className="s1c17x4y-0 jLdFhK s1xoa5y3-0 grmBGS"
  color="#66be54"
>
  + USD 10.00
</div>
`;

exports[`components CounterValue specifying ticker different from default 1`] = `
<div
  className="s1c17x4y-0 jLdFhK s1xoa5y3-0 grmBGS"
  color="#66be54"
>
  + USD 5.00
</div>
`;

exports[`components CounterValue using countervalue different from default 1`] = `
<div
  className="s1c17x4y-0 jLdFhK s1xoa5y3-0 bPBcmj"
  color={undefined}
>
  + EUR 0.42
</div>
`;

exports[`components CounterValue with time travel whith date in countervalues 1`] = `
<div
  className="s1c17x4y-0 jLdFhK s1xoa5y3-0 grmBGS"
  color="#66be54"
>
  + USD 20.00
</div>
`;

exports[`components CounterValue with time travel whith date not in countervalues 1`] = `
<div
  className="s1c17x4y-0 jLdFhK s1xoa5y3-0 grmBGS"
  color="#66be54"
>
  + USD 0.00
</div>
`;

exports[`components CounterValue without countervalues populated 1`] = `
<div
  className="s1c17x4y-0 jLdFhK s1xoa5y3-0 grmBGS"
  color="#66be54"
>
  + USD 0.00
</div>
`;
